File: neet_code_all_cpp/array_hash/optimal_partition_of_string.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

class Solution {
    public:
        // AC
        int partitionString(std::string s) {
            std::vector<std::pair<int,int>> result;
            std::unordered_set<char> set;
            int i = 0;
            while(i < s.size()) {
                int j = i;
                while(j < s.size() && !set.count(s[j])) {
                    set.insert(s[j]);
                    ++j;
                }
                result.push_back({i, j - 1});
                set.clear();
                i = j;
            }

            return (int)result.size();
        }

        // AC
        int partitionString2(std::string s) {
            int result = 0;
            std::unordered_set<char> set;
            int i = 0;
            while(i < s.size()) {
                int j = i;
                while(j < s.size() && !set.count(s[j])) {
                    set.insert(s[j]);
                    ++j;
                }
                ++result;
                set.clear();
                i = j;
            }

            return result;
        }
};

// Kotlinの模範解答より
class SolutionAnsKotlin {
    public:
        // AC
        int partitionString(std::string s) {
            int result = 0;
            std::unordered_set<char> set;

            for(auto& c: s){
                if(set.count(c)) {
                    ++result;
                    set.clear();
                }
                set.insert(c);
            }

            if(set.size() == 0) {
                return result;
            }
            return result + 1;
        }
};

int main(void) {
    std::string case_1 = "abacaba";
    // => 4
    std::string case_2 = "ssssss";
    // => 6

    Solution s_1;
    std::cout << s_1.partitionString(case_1) << std::endl;
    std::cout << s_1.partitionString(case_2) << std::endl;

    std::cout << s_1.partitionString2(case_1) << std::endl;
    std::cout << s_1.partitionString2(case_2) << std::endl;

    SolutionAnsKotlin s_ans_kt;
    std::cout << s_ans_kt.partitionString(case_1) << std::endl;
    std::cout << s_ans_kt.partitionString(case_2) << std::endl;
}

File: neet_code_all/src/array_hash/optimal_partition_of_string.rs
use std::collections::HashSet;

struct Solution {}
impl Solution {
    // AC
    pub fn partitoin_string(s: String) -> i32 {
        let s = s.chars().collect::<Vec<char>>();
        let mut i = 0;
        let mut result = vec![];
        let mut set = HashSet::new();
        while i < s.len() {
            let mut j = i;
            while j < s.len() && !set.contains(&s[j]) {
                set.insert(s[j]);
                j += 1;
            }
            //println!("i, j: {i},{j}");
            //println!("set: {:?}", set);
            result.push((i, j - 1));
            i = j;
            set.clear();
        }
        //println!("result: {:?}", result);

        result.len() as i32
    }

    // AC: こっちの方が省メモリ
    pub fn partitoin_string_2(s: String) -> i32 {
        let s = s.chars().collect::<Vec<char>>();
        let mut i = 0;
        let mut result = 0;
        let mut set = HashSet::new();
        while i < s.len() {
            let mut j = i;
            while j < s.len() && !set.contains(&s[j]) {
                set.insert(s[j]);
                j += 1;
            }
            //println!("i, j: {i},{j}");
            //println!("set: {:?}", set);
            result += 1;
            i = j;
            set.clear();
        }
        //println!("result: {:?}", result);

        result
    }
}

// Kotlinの模範解答より
struct SolutionAnsKotlin {}
impl SolutionAnsKotlin {
    // AC
    pub fn partitoin_string(s: String) -> i32 {
        let mut set = HashSet::new();
        let mut result = 0;

        for c in s.chars() {
            if set.contains(&c) {
                result += 1;
                set.clear();
            }
            set.insert(c);
        }

        if set.len() == 0 {
            return result;
        }

        result + 1
    }
}

fn main() {
    let case_1 = "abacaba".to_string();
    // => 4
    let case_2 = "ssssss".to_string();
    // => 6

    println!("case_1:{}", Solution::partitoin_string(case_1.clone()));
    println!("case_2:{}", Solution::partitoin_string(case_2.clone()));

    println!("case_1:{}", Solution::partitoin_string_2(case_1.clone()));
    println!("case_2:{}", Solution::partitoin_string_2(case_2.clone()));

    println!(
        "case_1:{}",
        SolutionAnsKotlin::partitoin_string(case_1.clone())
    );
    println!(
        "case_2:{}",
        SolutionAnsKotlin::partitoin_string(case_2.clone())
    );
}
